src/olympia/discovery/utils.py

```diff
@@ -41,17 +39,24 @@ def call_recommendation_server(id_or_guid, params, server):
 
 
 def get_recommendations(telemetry_id, params):
-    from olympia.addons.models import Addon  # circular import
+    from olympia.addons.models import Addon
+    from olympia.discovery.models import DiscoveryItem
     guids = call_recommendation_server(
         telemetry_id, params, settings.RECOMMENDATION_ENGINE_URL) or []
-    ids = (Addon.objects.public().filter(guid__in=guids)
-           .values_list('id', flat=True))
-    return [data.DiscoItem(addon_id=id_, is_recommendation=True)
-            for id_ in ids]
+    qs = Addon.objects.select_related('discoveryitem').public().filter(
+        guid__in=guids)
+    result = []
+    for addon in qs:
+        try:
+            addon.discoveryitem
+        except DiscoveryItem.DoesNotExist:
+            addon.discoveryitem = DiscoveryItem()
```

Comment:
Probably needs to be `DiscoveryItem(addon=addon)`. I'll add a comment and a test - basically this covers the case where a recommendation is coming from TAAR and we don't have a corresponding discovery item with custom strings for it. That's fine though, it will just build a heading and description from the add-on itself like it did before.

Reply:
yeah, `DiscoveryItem()` would just be an empty recommendation, right?

Reply:
Yeah, it's clearly wrong. Shows that a test is missing :)